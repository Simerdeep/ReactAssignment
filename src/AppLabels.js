export const MISSON_ID_LABEL = "Mission Ids";
export const LAUNCH_YEAR_LABEL = "Launch Year";
export const SUCCESSFULL_LAUNCH_LABEL = "Successfull Launch";
export const SUCCESSFULL_LANDING_LABEL = "Successfull Landing";
export const FILTERS_LABEL = "Filters";
export const SPACEXLAUNCH_PROGRAM_LABEL = "SpaceX Launch Programs";
export const FOOTER_LABEL = "Developed By: Simerdeep Singh";
export const NO_DATA_AVAILABLE = "No data available for particular filter";

export const SPACEX_API_URL = "https://api.spacexdata.com/v3/launches?limit=100";

