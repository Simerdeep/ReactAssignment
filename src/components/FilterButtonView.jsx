//@flow

import type {Element} from "React";
import React from "react";

type Props = {

    filterName: string,
    filterData: Array<string>,
    selectedFilter:? string,
    updateSelectedFilter: (filterValue:? string ) => void

};


/**
 * Component to show the buttons in the filter with the heading
 *
 * @param {Props} {
 *     filterName,
 *     filterData,
 *     selectedFilter,
 *     updateSelectedFilter
 * }
 * @returns {Element<"div">}
 */
const FilterButtonView = ({
    filterName,
    filterData,
    selectedFilter,
    updateSelectedFilter
}: Props ): Element<"div">  => {

    /**
     * To perform action when filter is clicked
     *
     * @param {*} filterValue selected filter value
     */
    const onFilterClick = (filterValue) => {

        if(selectedFilter === filterValue)
            updateSelectedFilter(null);
        else
            updateSelectedFilter(filterValue);
    };

    return (
        <div className="filter-button">

            <p className="align-center">{filterName}</p>
            <hr className="hr-width-20"/>

            <>

            {
                filterData.map((filterValue,index) => { 
                    
                    return(
                        <button 
                            className = {`button ${selectedFilter ? (filterValue.toString().toLowerCase() === selectedFilter.toString().toLowerCase() ? "active" : "") : ""}`} 
                            key = {index}
                            onClick = {() => onFilterClick(filterValue)}
                        >
                         {filterValue}
                        </button>
                    )
                    
                })
            }
            </>
        </div>

    );
}

export default FilterButtonView;
