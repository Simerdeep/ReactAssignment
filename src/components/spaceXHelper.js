//@flow

import { LAUNCH_YEAR_LABEL, MISSON_ID_LABEL, SUCCESSFULL_LANDING_LABEL, SUCCESSFULL_LAUNCH_LABEL } from "../AppLabels";

export const SUCCESSFULL_LAUNCH_VALUES = ["True","False"]; // Successfull launch default values
export const SUCCESSFULL_LANDING_VALUES = ["True","False"]; // Successfull landing default values

export const SEARCH_STRING_YEAR = "year"; // search string year to append in url 
export const SEARCH_STRING_LAUNCH = "launch"; // search string launch to append in url 
export const SEARCH_STRING_LANDING = "landing"; // search string landing to append in url 

const START_YEAR = 2006; // Start year for spacex year filter
const CURRENT_YEAR = new Date().getFullYear(); // Current Year

// Constant to hold the year values to show in the filter
export const YEAR_VALUES: Array<string> = Array.from({length: CURRENT_YEAR - START_YEAR}, (_, i) => (START_YEAR + i).toString());

// constant to hold the configuration to show for a spacex aircraft
export const spacexConfig = [

    {
        heading: "",
        accessor: (data: Object): any => data.mission_name + "#" + data.flight_number,
        customClass: "missionTitle",
        isNested: false
    },
    {
        heading: MISSON_ID_LABEL,
        accessor: "mission_id",
        isNested: true
    },
    {
        heading: LAUNCH_YEAR_LABEL,
        accessor: "launch_year",
        isNested: false
    },
    {
        heading: SUCCESSFULL_LAUNCH_LABEL,
        accessor: "launch_success",
        isNested: false
    },
    {
        heading: SUCCESSFULL_LANDING_LABEL,
        accessor: (data: Object): string => getLandingValue(data),
        isNested: false
    }
];

/**
 * To get the landing value of the spacex rocket
 *
 * @param {*} data data to use to fetch landing value
 * @returns landing value
 */
const getLandingValue = (data) => {

    if(data.rocket && data.rocket.first_stage && data.rocket.first_stage.cores && data.rocket.first_stage.cores.length > 0 )
        return data.rocket.first_stage.cores[0].land_success;

    return "";

}

/**
 * To append string 
 *
 * @param {*} stringValue current value of the string
 * @param {*} stringToAppendKey key to append in the string
 * @param {*} stringToAppendValue value of the string
 * @returns modified string
 */
const appendString = (stringValue,stringToAppendKey,stringToAppendValue) => {

    if(stringValue)
        return stringValue + `&${stringToAppendKey}=${stringToAppendValue}`;
    else
        return stringValue + `?${stringToAppendKey}=${stringToAppendValue}`; 

}

/**
 * To get search string based on the filters selected 
 *
 * @param {? string} year selected year
 * @param {?string} launch selected launch
 * @param {? string} landing selected landing
 * @returns {string} search string
 */
export const getSearchString = (year:? string, launch :?string,landing :? string ): string => {

    let searchString = "";

    if(year)
        searchString = appendString(searchString,SEARCH_STRING_YEAR,year);

    if(launch)
        searchString = appendString(searchString,SEARCH_STRING_LAUNCH,launch);

    if(landing)
        searchString = appendString(searchString,SEARCH_STRING_LANDING,landing);

    return searchString;

}

/**
 * To get the initial state of the filters
 *
 * @param {Object} history route history
 * @returns {{land_success: ?string, launch_success: ?string, launch_year: ?string}} return the default state values in state
 */
export const getInitialStateObject = (history: Object): {land_success: ?string, launch_success: ?string, launch_year: ?string} => {

    let initialStateObject = {
        launch_year: null,
        launch_success: null,
        land_success: null
    };

    if(history.location && history.location.search) {

        const searchValues = history.location.search.split("&");
    
        searchValues.forEach((searchValue) => {
    
            searchValue = searchValue.replace("?","");
            
            const [key,value] = searchValue.split("=");
            
            switch (key) {
                case SEARCH_STRING_YEAR:
                    initialStateObject.launch_year = value;
                    break;
        
                case SEARCH_STRING_LANDING:
                    initialStateObject.land_success = value.toString().toLowerCase();
                    break;
        
                case SEARCH_STRING_LAUNCH:
                    initialStateObject.launch_success = value.toString().toLowerCase();
                    break;
            }

            
        });
       
    
    }

    return initialStateObject;

}