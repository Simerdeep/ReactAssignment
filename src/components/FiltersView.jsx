//@flow

import type {Element} from "React";
import React, { useEffect, useState, useRef } from "react";
import { LAUNCH_YEAR_LABEL, SUCCESSFULL_LANDING_LABEL, SUCCESSFULL_LAUNCH_LABEL,FILTERS_LABEL  } from "../AppLabels";
import FilterButtonView from "./FilterButtonView";
import { SUCCESSFULL_LANDING_VALUES, SUCCESSFULL_LAUNCH_VALUES, YEAR_VALUES, getSearchString, getInitialStateObject } from "./spaceXHelper";
import { withRouter } from "react-router";
import "../css/FiltersView.css";

type Props = {
    history: Object,
    getSpaceXLaunchList: () => void
};

/**
 * Component to show the filters and handled state of filters
 *
 * @param {Props} props
 * @returns
 */
export const FiltersView = (props: Props): Element<"div"> => {

    const { getSpaceXLaunchList, history } = props;
    const initialStateObject = getInitialStateObject(history); // To get the initial state for the state variables
    
    const [selectedYear, updateSelectedYear] = useState(initialStateObject.launch_year); // To hold selected year filter
    const [successfullLaunch, updateSuccessfullLaunch] = useState(initialStateObject.launch_success); // To hold successfull launch filter
    const [successfullLanding, updateSuccessfullLanding] = useState(initialStateObject.land_success); // To hold successfull landing filter
    const isInitialMount = useRef(true); // To create a ref to check if component is getting mounted or update

    useEffect(() => {

        // If current value is true component is getting mounted then update it to false
        if (isInitialMount.current) {
            
            isInitialMount.current = false;
            
        } 

        // If component is getting update then update the route params and fetch new data based on filters
        else {

            let pathName = "/";

            if(selectedYear || successfullLaunch || successfullLanding)
                pathName = "/search";

            history.push({
                pathname: pathName,
                search   : getSearchString(selectedYear,successfullLaunch,successfullLanding)
            });

            getSpaceXLaunchList();
        }

    },[selectedYear,successfullLaunch,successfullLanding])

    return (
        <div className = "filters-view">

            <h3 className = "filter-heading">{FILTERS_LABEL}</h3>

            <FilterButtonView
                filterName={LAUNCH_YEAR_LABEL}
                filterData={YEAR_VALUES}
                selectedFilter={selectedYear}
                updateSelectedFilter={(year) => updateSelectedYear(year)}
            />

            <FilterButtonView
                filterName={SUCCESSFULL_LAUNCH_LABEL}
                filterData={SUCCESSFULL_LAUNCH_VALUES}
                selectedFilter={successfullLaunch}
                updateSelectedFilter={(launch) => updateSuccessfullLaunch(launch)}
            />

            <FilterButtonView
                filterName={SUCCESSFULL_LANDING_LABEL}
                filterData={SUCCESSFULL_LANDING_VALUES}
                selectedFilter={successfullLanding}
                updateSelectedFilter={(landing) => updateSuccessfullLanding(landing)}
            />

        </div>
        
    )

}

export default (withRouter(FiltersView) : any);