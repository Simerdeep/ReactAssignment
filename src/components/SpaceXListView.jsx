//@flow

import type {Element} from "React";
import React from "react";
import { spacexConfig } from "./spaceXHelper";
import "../css/SpaceXListView.css";
import { NO_DATA_AVAILABLE } from "../AppLabels"

type Props = {
    spaceXListData: Array<Object>
};

/**
 * Component to show the spacex list
 *
 * @param {Props} {
 *     spaceXListData
 * }
 * @returns {Element<"div">}
 */
const SpaceXListView = ({
    spaceXListData
}: Props): Element<"div"> => {


    /**
     * To render the nested view if required ( Required in case of mission ids)
     *
     * @param {*} data data to show in nested view
     * @param {*} config config to check for nested flag
     * @returns Element
     */  
    const getNestedView = (data,config) => {

        if(config.isNested){

            return (
                <ul>
                {
                    data[config.accessor].map((value,index) => <li key={index}>{value}</li>)
                }
                </ul>
            );

        }

        return null;

    }

    /**
     * To get the field value to show in the dom
     *
     * @param {*} data data to extract field from
     * @param {*} config config to extract acccessor from 
     * @returns value to be shown in dom
     */
    const getFieldValue = (data, config) => {

        const { accessor, isNested } = config;

        let value = "";

        if(typeof(accessor) === "function")
            value = accessor(data);
        else
            value = data[accessor];

        if(typeof(value) === "boolean")
            value = value.toString();

        if(!value || isNested)
            value = "";
        
        return value;

    }

    return (
        <div className="spaceXView">

            {

                spaceXListData.length > 0 ?
                <>
                {
                    spaceXListData.map((data,key) => {
                        
                        return (
                            <div className="card col-3" key={key}>

                                <div className="spaceXimagebox">
                                    <img src={data.links.mission_patch} alt="SpaceX" className="spaceXimage"/>
                                </div>
                                {
                                    spacexConfig.map((config,key) => {
                                        
                                        return(
                                            <div key={key}>
                                                {
                                                    config.heading ?
                                                    <p>
                                                        <span className="font-weight-bold">{config.heading}</span> :  
                                                        <span className="spacexValue">
                                                            { getFieldValue(data,config)}
                                                        </span>
                                                        
                                                    </p>
                                                    : <p className={config.customClass}>{getFieldValue(data,config)}</p>
                                                }
                                              
                                                { getNestedView(data,config) }
                                            </div>
                                            
                                        )
                                    })
                                }
        
                            </div>
        
                        )
                       
                    })
                }
                </>
                : <h1>{NO_DATA_AVAILABLE}</h1>
            }
        
        </div>
    );

}

export default SpaceXListView;