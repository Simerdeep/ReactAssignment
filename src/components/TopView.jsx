//@flow

import type {Node} from "React";
import Axios from "axios";
import React, { useEffect, useState } from "react";
import FiltersView from "./FiltersView";
import SpaceXListView from "./SpaceXListView";
import "../css/TopView.css";
import { withRouter } from "react-router";
import { getInitialStateObject } from "./spaceXHelper";
import { SPACEX_API_URL } from "../AppLabels";
import Loader from "./Loader";

type Props = {
    history: Object
};

/**
 * Component to show filters view and list of the spacex aircrafts
 *
 * @param {Props} props props received by the component
 * @returns Component
 */
export const TopView = (props: Props): Node => {

    // To hold list of spacex aircrafts
    const [spaceXListData, updateSpaceXListData] = useState([]); 
    const [isLoading,updateIsLoading] = useState(false);

    /**
     * To fetch spacex data by calling spacex api with the filters selected
     *
     */
    const getSpaceXLaunchList = async () => {

        let queryParameters = getInitialStateObject(props.history);
        updateIsLoading(true);

        try{
            const spaceXData = await Axios.get(SPACEX_API_URL,{params: queryParameters });
            updateSpaceXListData(spaceXData.data);
            updateIsLoading(false);
        }
        catch(e) {
            updateIsLoading(false);
        }
        
    }

    // To call spacex api whene component is mounted
    useEffect(() => { getSpaceXLaunchList() },[])

    return (
       <>

            <div className="col-2">
                <FiltersView 
                    getSpaceXLaunchList={getSpaceXLaunchList}
                />
            </div>

            <div className="col-10">
                {
                    isLoading ? 
                    <Loader />
                    :  <SpaceXListView 
                        spaceXListData={spaceXListData}
                    />
                }
               
            </div>
        
       </>
    )

};

export default (withRouter(TopView): any);