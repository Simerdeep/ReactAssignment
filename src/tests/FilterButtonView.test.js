import React from "react";
import { mount } from "enzyme";
import FilterButtonView from "../components/FilterButtonView";
import { LAUNCH_YEAR_LABEL } from "../AppLabels";
import { YEAR_VALUES } from "../components/spaceXHelper";


describe("Filter button view test cases",() => {

    const defaultProps = {
        filterName: LAUNCH_YEAR_LABEL,
        filterData: YEAR_VALUES,
        selectedFilter: null,
        updateSelectedFilter: jest.fn()
    };
    
    const filterButtonViewWrapper = mount(<FilterButtonView {...defaultProps}/>);

    it("should render the snapshot for filter button view", () => {
        expect(filterButtonViewWrapper).toMatchSnapshot();
    });

    it("should call on click method of button when same is not slected", () => {

        filterButtonViewWrapper.find("button").first().simulate("click");
        expect(filterButtonViewWrapper.props().updateSelectedFilter).toBeCalledWith(YEAR_VALUES[0]);
       
    });

    it("should call on click method of button when same button is already slected", () => {

        filterButtonViewWrapper.setProps({
            selectedFilter: YEAR_VALUES[0]
        });

        filterButtonViewWrapper.find("button").first().simulate("click");
        expect(filterButtonViewWrapper.props().updateSelectedFilter).toBeCalledWith(null);
    });

})