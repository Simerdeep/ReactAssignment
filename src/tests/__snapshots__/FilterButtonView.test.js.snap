// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter button view test cases should render the snapshot for filter button view 1`] = `ReactWrapper {}`;
